File: 2024/day18.py
import AOCInit
import util

import heapq as hq

if __name__ != '__main__':
    exit()

inp = util.getInput(d=18, y=2024)
blkSize = 70

w = blkSize + 1
h = blkSize + 1
goal = (blkSize, blkSize)

# part 1
count = 1024
corrupted = set(
    tuple(util.getInts(line))
    for line in inp.splitlines()[:count]
)
print(util.dijkstra(
        (0, 0),
        lambda nst, ost, oc: oc + 1,
        lambda ost: (pt for pt in util.nearby2DGridPts(ost, (w, h)) if pt not in corrupted),
        lambda st: st == goal
    )[1]
)

# part 2
corrupted = set()
for pt in (tuple(util.getInts(line)) for line in inp.splitlines()):
    corrupted.add(pt)
    buff = [(0, 0)]
    visited = set()
    while len(buff) != 0:
        x, y = buff.pop()
        if (x, y) in visited:
            continue
        visited.add((x, y))
        if (x, y) == goal:
            break
        for nb in util.nearby2DGridPts((x, y), (w, h)):
            if nb not in corrupted and nb not in visited:
                buff.append(nb)
    if goal not in visited:
        print(','.join(str(idx) for idx in pt))
        break

File: 2024/day18.cpp
#include <string>
#include <tuple>
#include <unordered_set>
#include <utility>
#include <vector>
#include <queue>

#include "../util.hpp"

constexpr int blkSize = 70 + 1;
constexpr int count = 1024;

using NodeState = std::tuple<int, int, int>;

bool isDisconnected(const std::unordered_set<int> &blocked){
    std::vector<int> buff(1, 0);
    std::unordered_set<int> visited;
    while (!buff.empty()){
        int idx = buff.back();
        buff.pop_back();
        if (util::contains(idx, visited)){
            continue;
        }
        visited.insert(idx);
        if (idx == blkSize * blkSize - 1){
            break;
        }
        for (const auto &nb : util::neighbourGridPoint(idx / blkSize, idx % blkSize, blkSize, blkSize)){
            int nbidx = nb.first * blkSize + nb.second;
            if (!util::contains(nbidx, blocked)){
                buff.push_back(nbidx);
            }
        }
    }
    return !util::contains(blkSize * blkSize - 1, visited);
}

int main(int, char**){
    std::vector<int> allCorrupted;
    for (auto &line : util::splitline(util::getInput(2024, 18))){
        auto v = util::extract_int(line);
        allCorrupted.push_back(v[0] * blkSize + v[1]);
    }

    // part 1
    std::queue<NodeState> q;
    std::unordered_set<int> corrupted(allCorrupted.cbegin(), allCorrupted.cbegin() + count), visited;
    q.push({0, 0, 0});
    while (!q.empty()){
        auto [cost, x, y] = q.front();
        q.pop();
        if (x == blkSize - 1 && y == blkSize - 1){
            util::output(cost);
            break;
        }
        if (util::contains(x * blkSize + y, visited)){
            continue;
        }
        visited.insert(x * blkSize + y);
        for (const auto &nb : util::neighbourGridPoint(x, y, blkSize, blkSize)){
            if (util::contains(nb.first * blkSize + nb.second, corrupted)){
                continue;
            }
            q.push({cost + 1, nb.first, nb.second});
        }
    }

    // part 2
    int s = 0, e = allCorrupted.size();
    while (s < e){
        int m = (s + e) / 2;
        if (isDisconnected(std::unordered_set<int>(allCorrupted.cbegin(), allCorrupted.cbegin() + m + 1))){
            e = m;
        } else {
            s = m + 1;
        }
    }
    util::printVec({allCorrupted[s] / blkSize, allCorrupted[s] % blkSize}, ",");

    return 0;
}
